File: ASSIGNMENT_ARules_groceries.R
################### Assignment ARules on groceries dataset

install.packages('arules')
library(arules)
library(arulesViz) # for visualizing 

data('Groceries')
class(Groceries)
str(Groceries)

summary(Groceries)
# 9835 transactions have been done by customers
# the total products (or items) are 169
# top product purchased is whole milk. It has been purchased in 2513 transactions
# In 2519 transactions only one product has been purchased. In 3 transactions each
# 29 items has been purchased.
# minimum one item has been purchaded and maximum 32 items have been purchased
# per transaction.

inspect(Groceries[1:4]) # viewing first 4 transactions

# visualizing top20 transactions
itemFrequencyPlot(Groceries,topN=20)

# let us now build rules using apriori package
arules <- apriori(Groceries, parameter=list(support=0.002,confidence=0.6,minlen=2))
# support=0.002 means atleast 20 times in entire 10,000(9835) trans it should be present
arules # set of 376 rules 
inspect(head(sort(arules,by='lift')))
# get top 6 rules as per highest lift

head(quality(arules))
inspect(head(arules))

# visualize
plot(arules)
# when support decreases conf also decreases. 

# keep changing support and lift
arules1 <- apriori(Groceries,parameter=list(support=0.002,confidence=0.7,minlen=3))
# for minlen=3, get 94 rules
# for minlen=4, get 8 rules
# for minlen=5, get 5 rules
arules1a <- apriori(Groceries,parameter = list(support=0.002,confidence=0.75,minlen=3))
# for conf=0.7, get 94 rules
# for conf=0.75, get 39 rules
# for conf=0.8, get 11 rules
arules3 <- apriori(Groceries,parameter=list(support=0.003,confidence=0.7))
# 19 rules
arules3[is.redundant(arules3)] # 0 rules
inspect(arules3)

arules2 <- apriori(Groceries,parameter = list(support=0.002,confidence=0.8,minlen=3))
inspect(arules2)
inspect(sort(arules2,by='lift'))
plot(arules2)
# have 2 rules with high lift
plot(arules2,method = 'grouped')
plot(arules2,method = 'graph')

arules2_df = data.frame(
  lhs = labels(lhs(arules2)),
  rhs = labels(rhs(arules2)), 
  arules2@quality)

write.csv(arules2_df,file='E:\\Association_Rules\\arules2.csv')
# unlink('E:\\Association_Rules\\arules2a.csv')


# removing redundant rules
arules2[is.redundant(arules2)]
# set of 0 rules, there are no redundant rules

# arules2_redn <- arules2[!is.redundant(arules2)]
inspect(arules2_redn)

arules # set of 376 rules 
arules[is.redundant(arules)] # there are 6 redundant rules

# library(readxl)
library(writexl)
write_xlsx(arules2_df,'E:\\Association_Rules\\arules2.xlsx')

File: ASSIGNMENT_ARules_groceries.py
# ASSIGNMENT ASSOCIATION_RULES groceries dataset

# importing important libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from mlxtend.frequent_patterns import apriori, association_rules


groceries=[]
with open('E:/Datasets/groceries.csv') as f:
    groceries = f.read()
groceries=groceries.split('\n')    
groceries_list=[]
for i in groceries:
    groceries_list.append(i.split(','))

groceries_df = pd.DataFrame(pd.Series(groceries_list))
groceries_df.columns=['Transactions']
groceries_df[9833:]
groceries_df = groceries_df.iloc[:9835,:]
X = groceries_df.Transactions.str.join(sep='*').str.get_dummies(sep='*')

# getting a list of frequent items with atleast 3% frequency in the dataset
frequent_itemsets = apriori(X,  min_support=0.03,use_colnames=True)
frequent_itemsets.shape # 64 
frequent_itemsets.sort_values('support',ascending=False).head(5)
#Quick look at the top 25 items
X.sum().to_frame('Frequency').sort_values('Frequency',ascending=False)[:25].plot(kind='bar',
                                                                                 figsize=(12,8),title="Frequent Items");plt.show();

''' we have 64 items that appear atleast 3% of the times in the groceries
dataset. whole milk has highest support of 0.25 i.e., it appears 
25% times in the dataset, one-fourth of the transactions are made of
whole milk: followed by other-vegetables which appear 19% of the times.
'''

# Getting rules from the frequent itemsets
rules=association_rules(frequent_itemsets, metric='lift',min_threshold=1)
rules.shape # we have obtained 34 rules 
rules.head()

############# removing redundant rules
'''
we find that rules 0 and 1 are duplicate rules. similarly rule 2 and 3 are same
rules. we have to remove such redundant rules
'''
def to_list(i):
    return (sorted(list(i)))
ma_X = rules.antecedents.apply(to_list)+rules.consequents.apply(to_list)
ma_X = ma_X.apply(sorted)
rules_sets = list(ma_X)
unique_rules_sets = [list(m) for m in set(tuple(i) for i in rules_sets)]
index_rules = []
for i in unique_rules_sets:
    index_rules.append(rules_sets.index(i))


# getting rules without any redudancy 
rules_no_redundancy  = rules.iloc[index_rules,:]

# Sorting them with respect to lift and getting top 10 rules 
rules_no_redundancy.sort_values('lift',ascending=False).head(10)
rules.shape # we have 34 rules including duplicate rules
rules_no_redundancy.shape # now we have 17 rules after removing duplicate rules
rules_no_redundancy.sort_values('lift',ascending=False)
'''
The top-most rule is root-vegetables and other-vegetables. this rule appears
4.7% of the times in the whole transactions. when a person purchases root-vegs,
we can be 43% confident that he will also purchase other-vegs. 
If Lift>1 indicates positive dependence or complementary effect. 
The lift valus is 2.25 indicating that they are positively correlated.

'''
# visualization
%matplotlib inline
# support vs confidence
plt.scatter(rules_no_redundancy.support, rules_no_redundancy.confidence, marker='*') 
plt.xlabel('support'); plt.ylabel('confidence'); plt.title('Support vs Confidence')

# support vs lift
plt.scatter(rules_no_redundancy.support, rules_no_redundancy.lift, marker='*') 
plt.xlabel('support'); plt.ylabel('lift'); plt.title('Support vs Lift')

# lift vs confidence
plt.scatter(rules_no_redundancy.lift, rules_no_redundancy.confidence)
# positive correlation is visible from the scatter plot
fit = np.polyfit(rules_no_redundancy['lift'], rules_no_redundancy['confidence'], 1)
fit_fn = np.poly1d(fit)
plt.plot(rules_no_redundancy['lift'], rules_no_redundancy['confidence'], 'yo', rules_no_redundancy['lift'], 
 fit_fn(rules_no_redundancy['lift']))
plt.xlabel('lift');plt.ylabel('confidence');plt.title('Lift vs Confidence');plt.show();

#### RECOMMENDATIONS
'''
1. those who buy root-vegs we can recommend to them other-vegs, whole-milk,fruits
2. Looking at all 17 rules, we see that root-vegs, other-vegs, whole-milk,
    rolls/buns, fruits(tropical and pip fruit),soda, bottled water, yogurt 
    have lift>1.19
    we can place these items in side-by-side racks.

'''

